### PR TITLE
Set cluster updated cond to unknown once node is added/deleted

### PR DESCRIPTION
**Pull request by zionwu**
https://github.com/rancher/rancher/issues/11152

### CODE REVIEW

File: pkg/controllers/management/clusterprovisioner/provisioner.go [hunk 1]
```diff
@@ -106,6 +106,12 @@ func (p *Provisioner) Updated(cluster *v3.Cluster) (*v3.Cluster, error) {
 	obj, err := v3.ClusterConditionUpdated.Do(cluster, func() (runtime.Object, error) {
 		return p.update(cluster, false)
 	})
+
+	changed, err := p.nodeChanged(obj.(*v3.Cluster))
+	if err == nil && changed {
+		v3.ClusterConditionUpdated.Unknown(obj)
```
Comment (alena1108) [inline pkg/controllers/management/clusterprovisioner/provisioner.go]: @zionwu Could you point the place where ClusterConditionUpdated gets reset *from* unknown, back to ready?

Comment (zionwu) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: In the same func https://github.com/rancher/rancher/pull/12118/files#diff-6d27775f3dc5a957254ede85cdb01c10L107:
```
obj, err := v3.ClusterConditionUpdated.Do(cluster, func() (runtime.Object, error) {
 	return p.update(cluster, false)
 })
```
When the node finish provisioned, the above code will change the `ClusterConditionUpdated` to `True`.  Func `nodeChange` will return `false`  and it will not change the state to `unknown`.

Comment (alena1108) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: @zionwu unknown usually indicates transitioning. So cluster goes to Unknown as soon as this block is entered: https://github.com/rancher/rancher/pull/12118/files#diff-6d27775f3dc5a957254ede85cdb01c10R106. So it goes True->Unknown->True at line 107 if the update is successful. If update fails, it goes to False.

The fact that it would need get back to Unknown at line 112 indicates some disagreement with results returned by p.Update. 

Also nodeChanged code seems to be very similar to 
getConfig-> reconcileRKENodes() call, which is being done when p.update is being called. So when node is not ready, the cluster would be marked as Provisioned=false anyway.

Could you double check the code, and see if it would be enough just to remove the https://github.com/rancher/rancher/pull/12118/files#diff-6d27775f3dc5a957254ede85cdb01c10L124 condition, so cluster update is being called when the node is not fully ready, so https://github.com/rancher/rancher/pull/12118/files#diff-6d27775f3dc5a957254ede85cdb01c10R107 can set the Provisioned condition to false, and the call would be retried

Comment (zionwu) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: @alena1108 just remove the condition https://github.com/rancher/rancher/pull/12118/files#diff-6d27775f3dc5a957254ede85cdb01c10L124 does not work, I have tried and did the testing.  That is why I have to add more code to fix the issue.   

In the reconcileCluster func, before doing actual work, there is logic to check the spec, https://github.com/rancher/rancher/pull/12118/files#diff-6d27775f3dc5a957254ede85cdb01c10R230:
```
spec, err := p.getSpec(cluster)
 	if err != nil || spec == nil {
 		return cluster, err
 	}
```
` p.getSpec` will call p.getConfig to compare the latest config and applied config.  If you take a look at the `getConfig` func, you will find the latest config will not take the node being provisioned/removed into account.  So `getSpec` will return a nil when a node is just added/removed.   From the above code snippet, if spec is nil, it will return  the cluster obj with nil error.  So the ClusterConditionUpdated and ClusterConditionProvisioned will always be true for the case that a node is just added but not finish provisioning.

Comment (alena1108) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: @zionwu ok..my only concern then is we are setting the condition to Unknown manually, relying on  external controllers to trigger the cluster update when the node gets ready. Instead of returning the error and resetting the condition to false (using v3.ClusterConditionUpdated.Do callback), which would in turn reschedule cluster update call on the spot. If you are comfortable, and 100% positive that the cluster never gets stuck in Unknown state with this implementation, I'l merge :)

Comment (zionwu) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: @alena1108 , I checked the `ClusterConditionUpdated.Do`, its logic will set ClusterConditionUpdated to false if the callback func return err.  I also want to add the logic inside the `update` func, but I can't find a way to let the status stays Unknown as `ClusterConditionUpdated.Do` set it to true or false.

Comment (zionwu) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: The reason why ClusterConditionUpdated need to be `Unknown` is that UI will display the state as `updating` only when it is `Unknown`

Comment (alena1108) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: `Do you think we should make change to ClusterConditionUpdated.Do so that the cond can be set to Unknown` - Unknown means transitioning; framework can't set it to be a final state. 

I guess the way we set it in PR to unknown explicitly, would be the only way to make it stay in Unknown state w/o holding up the thread, to satisfy`The reason why ClusterConditionUpdated need to be Unknown is that UI will display the state as updating only when it is Unknown`. I just want to make sure that the condition gets reset eventually, could you test all corner cases scenarios?

Comment (zionwu) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: Okay, I will spend more time to test all corner cases.

Comment (zionwu) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: @alena1108  I tested with the following cases:
1. add a new node pool to an active cluster
2. remove a new node pool to an active cluster
3. increase the node count for a node pool
4. decrease the node count for a node poll
5. delete a node in the node pool

All cases passed.  Am I missing any cases that you cared about?

Comment (alena1108) [reply pkg/controllers/management/clusterprovisioner/provisioner.go]: @zionwu that should be good enough
